README.org
#+TITLE: Mi configuración de Emacs
#+AUTHOR: Vim y LATEX en español
#+OPTIONS: num:nil 
#+LANGUAGE: es

* Índice :toc:
- [[#introducción][Introducción]]
- [[#ajustes-iniciales][Ajustes iniciales.]]
- [[#ajustes-visuales][Ajustes visuales]]
- [[#atajo-de-teclado-para-org-agenda][Atajo de teclado para ~org-agenda~]]
- [[#para-mostrar-números-de-línea-relativo-como-en-vim][Para mostrar números de línea relativo (como en VIM)]]
- [[#eliminando-los-números-de-línea-en-modos-de-edición-de-texto-solo-los-veremos-en-latex-y-en-programación][Eliminando los números de línea en modos de edición de texto (solo los veremos en LATEX y en programación)]]
- [[#configurando-navegador][Configurando navegador]]
- [[#configuración-de-paquetes][Configuración de paquetes]]
  - [[#agregaremos-melpa][Agregaremos melpa]]
  - [[#markdown-mode-requiere-la-instalación-de-use-package][markdown-mode (requiere la instalación de use-package)]]
  - [[#emmet-mode][Emmet mode]]
  - [[#darkroom-mode][Darkroom mode.]]
  - [[#toc-org][toc-org]]
  - [[#org-superstar-e-indentación-de-archivos-de-org-mode][Org-superstar e indentación de archivos de ~org-mode~]]
  - [[#lorem-ipsum-3][Lorem-ipsum :3]]
  - [[#counsel-y-swiper-para-búsqueda][Counsel y Swiper para búsqueda]]
  - [[#doom-modeline-la-barrita-de-doom-emacs][doom-modeline: La barrita de Doom Emacs]]
  - [[#emms][EMMS]]
  - [[#move-text][Move-text]]
- [[#previsualización-de-latex-en-org-mode][Previsualización de LATEX en org-mode]]
- [[#para-ver-lo-que-vienen-siendo-los-emojis][Para ver lo que vienen siendo los emojis]]
- [[#configuración-de-org-capture][Configuración de org-capture]]
- [[#para-evitar-transparencias-no-es-estrictamente-necesario][Para evitar transparencias (no es estrictamente necesario)]]

* Introducción

Emacs es un poderoso intérprete Lisp que tiene la característica de editar texto. Emacs es tan personalizable como lo requiramos. Y en este caso, esta configuración se acopla a lo que necesito de emacs como editor.

Si quieren usar esta configuración, lo recomendable es que revisen personalmente qué es lo que les sirve de ella, y lo copien a su init.el. Si lo que requieren es probar mi propia configuración, adelante. Pueden clonar este repositorio, con la advertencia de que:

1. Está configurado dentro de Linux Mint. En sistemas basados en Arch, la configuración de algunos plugins como emms cambian. Y tal vez hable de ello cuando hable en específico de Emms.
2. Algunos de los paquetes llevan modificaciones que no están implementadas directamente en los repositorios de los paquetes instalados. Proceder con precaución.
3. Dentro de init.el, en el apartado de ~package-selected-packages~ se encuentra la lista de paquetes que se han instalado desde ~M-x package-install RET~ (o sea, dentro de emacs). Instalarlos de esa manera los almacena dentro de la carpeta ~elpa~ que se encuentra casi siempre en ~/home/<nombre de usuario>/.emacs.d/~
4. Requiere la tipografía "Noto Color Emoji" que nos va a servir para poder visualizar emojis en emacs (No es forzoso. Si no lo requieren, simplemente eliminen todo lo referente a emojify de este mismo archivo de configuración.

* Ajustes iniciales.

Para poder usar emacs con esta configuración deberás crear un archivo llamado 'init.el', el cual deberá contener lo siguiente:

#+begin_src

(org-babel-load-file
 (expand-file-name
  "config.org"
  user-emacs-directory))

#+end_src

De esta manera, podemos ir agregando nuestra configuración en el archivo 'config.org'

* Ajustes visuales

Vamos a:

1. Pedir que emacs inicie maximizado.
2. Quitar el mensaje de bienvenida.
3. Esconder la barra de desplazamiento.
4. Quitar la barra de herramientas.
5. Vamos a agrandar nuestro "margen".
6. Quitar la barra de menú.

#+begin_src emacs-lisp

(add-hook 'window-setup-hook 'toggle-frame-maximized t)
(setq inhibit-startup-message t)
(scroll-bar-mode -1)
(tool-bar-mode -1)
(set-fringe-mode 10)
(menu-bar-mode -1)

#+end_src

Esto es para mostrar el tiempo de carga de emacs.

#+begin_src emacs-lisp
;; Startup timer
(add-hook 'emacs-startup-hook
          (lambda ()
            (message "Emacs ready in %s with %d garbage collections."
                     (format "%.2f seconds"
                             (float-time
                              (time-subtract after-init-time before-init-time)))
                     gcs-done)))
#+end_src

A continuación vamos a cambiar la fuente y el tema
(Debemos instalar Fira Code Retina previamente)

Siempre podemos agregar temas distintos de manera manual.

#+begin_src emacs-lisp

;;(set-face-attribute 'default nil :font "Fira Code Retina" :height 110)
(set-face-attribute 'default nil
:font "IBM Plex Mono"
:height 110)

;;(load-theme 'tango-dark)
;;(load-theme 'solarized-dark t)
(load-theme 'spacemacs-dark t)
(setq visible-bell t)
#+end_src

Quitemos el molesto doble espaciado para identificar el final de una oración/enunciado. Yo nunca lo he utilizado.

#+begin_src emacs-lisp

(setq sentence-end-double-space nil)

#+end_src

* Atajo de teclado para ~org-agenda~
#+begin_src emacs-lisp

(global-set-key (kbd "C-c a") 'org-agenda)

#+end_src

* Para mostrar números de línea relativo (como en VIM)

Números de línea relativo. Nos van a indicar el número de renglones por encima y por debajo del cursor actual. Y nos muestra el número de línea ocupado por el cursor.

He visto muchas configuraciones que insisten en instalar un paquete llamado linum. No es necesario agregar un paquete extra a Emacs para poder mostrar los números de línea.

#+begin_src emacs-lisp

(setq display-line-numbers-type 'relative)
(global-display-line-numbers-mode)
(global-visual-line-mode t)

#+end_src

* Eliminando los números de línea en modos de edición de texto (solo los veremos en LATEX y en programación)

Declaramos una función que elimina los números de línea, y la invocamos con la carga de ciertos modos, como son:
- org-mode
- markdown-mode
- eshell-mode (nadie quiere números de línea con su terminal)
- dired-mode (supongo que nadie quiere números de línea en su administrador de archivos)

#+begin_src emacs-lisp

(defun no-despliegues ()
  (display-line-numbers-mode -1)
)
(add-hook 'org-mode-hook 'no-despliegues)
;;(add-hook 'text-mode-hook 'no-despliegues)
(add-hook 'markdown-mode-hook 'no-despliegues)
(add-hook 'eshell-mode-hook 'no-despliegues)
(add-hook 'dired-mode-hook 'no-despliegues)

#+end_src

* Configurando navegador

Configuración un tanto innecesaria, pero puede funcionar para forzar a emacs a abrir el navegador firefox para mostrar los sitios web. En mi caso, casi siempre me abría un navegador web que no es el principal antes de agregarle ésta línea.

#+begin_src emacs-lisp

;; El navegador por default es firefox
(setq browse-url-browser-function 'browse-url-firefox)

#+end_src

* Configuración de paquetes

** Agregaremos melpa

Melpa es un repositorio de paquetes para emacs del que se pueden descargar paquetes bastante útiles. Se agregan también elpa y org.

#+begin_src emacs-lisp

(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("org" . "https://orgmode.org/elpa/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))


#+end_src

** markdown-mode (requiere la instalación de use-package)

Markdown es un lenguaje de marcado universal. "Markdown mode" facilita la edición de archivos de markdown. He de ser honesto que para convertirlo a otro formato, prefiero usar pandoc desde eshell, por lo cual, ~init~ y ~bind~ no son tan necesarios. Además de ser comandos que no uso.

#+begin_src emacs-lisp

(use-package markdown-mode
  :ensure t
  :mode ("README\\.md\\'" . gfm-mode))
  ;;:init (setq markdown-command "multimarkdown")
  ;;:bind (:map markdown-mode-map
         ;;("C-c C-e" . markdown-do)))

#+end_src

** Emmet mode

Para la sintaxis de nuestros queridos documentos html. Emmet está basado en el plugin del mismo nombre del editor de texto VSCode con el que se pueden editar documentos HTML con una sintaxis muy simple.

En este caso coloco el archivo de emmet-mode.el en la subcarpeta lisp, y le pido a emacs que añada esta carpeta a la ruta de carga.

Cabe destacar que este archivo está modificado por una cuestión que hay todavía con la librería ~cl~.

#+begin_src emacs-lisp

(add-to-list 'load-path "~/.emacs.d/lisp")
(require 'emmet-mode)

(add-hook 'sgml-mode-hook 'emmet-mode)
(add-hook 'css-mode-hook 'emmet-mode)

#+end_src

** Darkroom mode.

Para poder editar en un modo libre de distracciones. Es bastante parecido a Goyo en VIM. Otro plugin que hace lo mismo es Olivetti (que también tengo instalado).

#+begin_src emacs-lisp

(require 'darkroom)

#+end_src

** toc-org

Para crear una tabla de contenidos dentro de un documento de org-mode. Para invocarlo se requiere usar la etiqueta :toc: en un encabezado o subencabezado.

#+begin_src emacs-lisp

(add-to-list 'load-path "~/.emacs.d/lisp/toc-org")
(if (require 'toc-org nil t)
    (progn
      (add-hook 'org-mode-hook 'toc-org-mode))
      ;; enable in markdown, too
      ;;(add-hook 'markdown-mode-hook 'toc-org-mode)
      ;;(define-key markdown-mode-map (kbd "\C-c\C-o") 'toc-org-markdown-follow-thing-at-point))
  (warn "toc-org not found"))

#+end_src

** Org-superstar e indentación de archivos de ~org-mode~

Sucesor de org-bullets. Sirve para que org-mode se vea más bonito. Comentado para que no sea forzoso.

Lo que sí queda activada es la indentación en Org-mode.

#+begin_src emacs-lisp

;;(require 'org-superstar) <- En caso de no tenerlo instalado.
(add-hook 'org-mode-hook (lambda () (org-superstar-mode 1)))

(setq org-startup-indented t)

#+end_src

** Lorem-ipsum :3

Para agregar texto lorem ipsum.

#+begin_src emacs-lisp

(require 'lorem-ipsum)

#+end_src
** Counsel y Swiper para búsqueda

Esta configuración facilita la búsqueda de palabras y de archivos.

#+begin_src emacs-lisp

(use-package counsel
  :ensure t
)
(use-package swiper
  :ensure try
  :config
  (progn
  (ivy-mode)
  (setq ivy-use-virtual-buffers t)
  (setq enable-recursive-minibuffers t)
  ;; enable this if you want `swiper' to use it
  ;; (setq search-default-mode #'char-fold-to-regexp)
  (global-set-key "\C-s" 'swiper)
  (global-set-key (kbd "C-c C-r") 'ivy-resume)
  (global-set-key (kbd "<f6>") 'ivy-resume)
  (global-set-key (kbd "M-x") 'counsel-M-x)
  (global-set-key (kbd "C-x C-f") 'counsel-find-file)
  (global-set-key (kbd "<f1> f") 'counsel-describe-function)
  (global-set-key (kbd "<f1> v") 'counsel-describe-variable)
  (global-set-key (kbd "<f1> o") 'counsel-describe-symbol)
  (global-set-key (kbd "<f1> l") 'counsel-find-library)
  (global-set-key (kbd "<f2> i") 'counsel-info-lookup-symbol)
  (global-set-key (kbd "<f2> u") 'counsel-unicode-char)
  (global-set-key (kbd "C-c g") 'counsel-git)
  (global-set-key (kbd "C-c j") 'counsel-git-grep)
  (global-set-key (kbd "C-c k") 'counsel-ag)
  (global-set-key (kbd "C-x l") 'counsel-locate)
  (global-set-key (kbd "C-S-o") 'counsel-rhythmbox)
  (define-key minibuffer-local-map (kbd "C-r") 'counsel-minibuffer-history)
  ))

#+end_src

** doom-modeline: La barrita de Doom Emacs

Es una elegante barrita que nos da la información necesaria sobre nuestro archivo. 

#+begin_src emacs-lisp

(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1)
  :custom ((doom-modeline-height 15)))

#+end_src

** EMMS

Para reproducir audio dentro de Emacs. Esta configuración solo me sirve en linux mint. Para usarlo en sistemas basados en arch, solo hay que hacer unas modificaciones que agregaré más adelante.

#+begin_src emacs-lisp

(add-to-list 'load-path "~/.emacs.d/lisp/emms")
(require 'emms-setup)
  (emms-all)
  (emms-default-players)
  (setq emms-source-file-default-directory "~/Music/") ;; Change to your music folder

#+end_src

** Move-text

Para poder inicializar move-text, que nos sirve para mover líneas con M-up M-down. Instalarla con package-install

#+begin_src emacs-lisp

(move-text-default-bindings)

#+end_src

* Previsualización de LATEX en org-mode

Con el comando C-c C-x C-l se puede mostrar ecuaciones dentro de nuestro buffer actual de org-mode. Muy útil para apuntes.

#+begin_src emacs-lisp
(setq org-format-latex-options (plist-put org-format-latex-options :scale 2.0))
#+end_src

* Para ver lo que vienen siendo los emojis

Requiere la instalación de "Noto Color Emoji". Una fuente que puede desplegar los emoji en Emacs. Por otra parte, emojify permite insertar emojis dentro de nuestro bufer. Tal vez no sea la función más útil, pero puede servir para algo eventualmente.

#+begin_src emacs-lisp

(set-fontset-font t 'symbol (font-spec :family "Noto Color Emoji") nil 'prepend)
(use-package emojify
  :config
  (when (member "Noto Color Emoji" (font-family-list))
    (set-fontset-font
     t 'symbol (font-spec :family "Noto Color Emoji") nil 'prepend))
  (setq emojify-display-style 'unicode)
  (setq emojify-emoji-styles '(unicode))
  (bind-key* (kbd "C-c .") #'emojify-insert-emoji)) ; override binding in any mode

#+end_src

* Configuración de org-capture

Org-capture es un modo que permite tomar notas sobre las mismas notas.

#+begin_src emacs-lisp
(setq org-capture-templates
      '(("t" "Tarea" entry (file+headline "~/Documentos/org-mode/tareas.org" "Tareas")
         "* TODO %?\n  %i\n  %a")
        ("n" "Nota" entry (file+headline "~/Documentos/org-mode/notas.org" "Notas")
         "* %?\n  %i\n  %a")))

(global-set-key (kbd "C-c c") 'org-capture)
#+end_src

* Para evitar transparencias (no es estrictamente necesario)

#+begin_src emacs-lisp
(set-frame-parameter (selected-frame) 'alpha '(100 . 100))
 (add-to-list 'default-frame-alist '(alpha . (100 . 100)))
#+end_src


eshell/history
gcc -o 
gcc -o numerosSuma numerosSuma.c
./numerosSuma
./numerosSuma
gcc -o numerosSuma numerosSuma.c
./numerosSuma
gcc -o euclides euclides.c
./euclides
./euclides
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
./euclides
gcc -o euclides euclides.c
./euclides
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
gcc -o euclides euclides.c
./euclides
gcc -o euclides euclides.c
./euclides
gcc -o hola hola.cpp
g++ -o hola hola.cpp
./hola
yt-dlp -f 251 https://www.youtube.com/watch?v=QXQQAsIhHMw
ffmpeg -i ROSALÍA,\ Rauw\ Alejandro\ -\ BESO\ \(Official\ Video\)\ \[QXQQAsIhHMw\].webm -b:a 128k Beso.ogg
exit
gcc -o hola hola.c
gcc -o hola hola.c
./hola
gcc -o hola hola.c
./hola
gcc -o hola hola.c
./hola
./hola
gcc -o hola hola.c
./hola
exit
exit
sent pres.txt
exit
make clean
sudo make install
exit
sent pres.txt7
sent pres.txt
cd sent/
nano config.h
make clean
sudo make install
exit
sent pres.txt
sent pres.txt
sent pres.txt
ls  
reboot
sent pres.txt
sent pres.txt
sent pres.txt
sent pres.txt
sent pres.txt
sent pres.txt
sent pres.txt
sent pres.txt
sent pres.txt
sent pres.txt
sent pres.txt
du -sh pres.txt
exi
exit
du -sh auto-save-list/
cd auto-save-list/
ls
cd ..
ls
rm bookmarks
rm network-security.data
rm recentf*
exit
cd .emacs.d/
ls -la
exit
cd .emacs.d/
rm README.org
ln config.org README.org
ls -la 
rm config.html
exit
gcc -o quicksort quicksort.c
./quicksort
gcc -o quicksort quicksort.c
./quicksort
gcc -o quicksort quicksort.c
./quicksort
exit
mpv https://www.instagram.com/karen.psicologiacongracia/live/
exit
yt-dlp --cookies-from-browser firefox https://www.instagram.com/stories/mileyl026/3115057951870806599/
mpv Video\ by\ None\ \[Cs66LP0A-pH\].mp4
mpv Video\ by\ None\ \[Cs66LP0A-pH\].mp4
exit


config.el
(add-hook 'window-setup-hook 'toggle-frame-maximized t)
(setq inhibit-startup-message t)
(scroll-bar-mode -1)
(tool-bar-mode -1)
(set-fringe-mode 10)
(menu-bar-mode -1)

;; Startup timer
(add-hook 'emacs-startup-hook
          (lambda ()
            (message "Emacs ready in %s with %d garbage collections."
                     (format "%.2f seconds"
                             (float-time
                              (time-subtract after-init-time before-init-time)))
                     gcs-done)))

;;(set-face-attribute 'default nil :font "Fira Code Retina" :height 110)
(set-face-attribute 'default nil
:font "IBM Plex Mono"
:height 110)

;;(load-theme 'tango-dark)
;;(load-theme 'solarized-dark t)
(load-theme 'spacemacs-dark t)
(setq visible-bell t)

(setq sentence-end-double-space nil)

(add-to-list 'default-frame-alist '(fullscreen . maximized))

(global-set-key (kbd "C-c a") 'org-agenda)

(setq display-line-numbers-type 'relative)
(global-display-line-numbers-mode)
(global-visual-line-mode t)

(defun no-despliegues ()
  (display-line-numbers-mode -1)
)
(add-hook 'org-mode-hook 'no-despliegues)
;;(add-hook 'text-mode-hook 'no-despliegues)
(add-hook 'markdown-mode-hook 'no-despliegues)
(add-hook 'eshell-mode-hook 'no-despliegues)
(add-hook 'dired-mode-hook 'no-despliegues)

;; El navegador por default es firefox
(setq browse-url-browser-function 'browse-url-firefox)

(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("org" . "https://orgmode.org/elpa/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))

(use-package markdown-mode
  :ensure t
  :mode ("README\\.md\\'" . gfm-mode))
  ;;:init (setq markdown-command "multimarkdown")
  ;;:bind (:map markdown-mode-map
         ;;("C-c C-e" . markdown-do)))

(add-to-list 'load-path "~/.emacs.d/lisp")
(require 'emmet-mode)

(add-hook 'sgml-mode-hook 'emmet-mode)
(add-hook 'css-mode-hook 'emmet-mode)

(require 'darkroom)

(add-to-list 'load-path "~/.emacs.d/lisp/toc-org")
(if (require 'toc-org nil t)
    (progn
      (add-hook 'org-mode-hook 'toc-org-mode))
      ;; enable in markdown, too
      ;;(add-hook 'markdown-mode-hook 'toc-org-mode)
      ;;(define-key markdown-mode-map (kbd "\C-c\C-o") 'toc-org-markdown-follow-thing-at-point))
  (warn "toc-org not found"))

;;(require 'org-superstar) <- En caso de no tenerlo instalado.
;;(add-hook 'org-mode-hook (lambda () (org-superstar-mode 1)))

;;(setq org-startup-indented t)

(require 'lorem-ipsum)

(use-package counsel
  :ensure t
)
(use-package swiper
  :ensure try
  :config
  (progn
  (ivy-mode)
  (setq ivy-use-virtual-buffers t)
  (setq enable-recursive-minibuffers t)
  ;; enable this if you want `swiper' to use it
  ;; (setq search-default-mode #'char-fold-to-regexp)
  (global-set-key "\C-s" 'swiper)
  (global-set-key (kbd "C-c C-r") 'ivy-resume)
  (global-set-key (kbd "<f6>") 'ivy-resume)
  (global-set-key (kbd "M-x") 'counsel-M-x)
  (global-set-key (kbd "C-x C-f") 'counsel-find-file)
  (global-set-key (kbd "<f1> f") 'counsel-describe-function)
  (global-set-key (kbd "<f1> v") 'counsel-describe-variable)
  (global-set-key (kbd "<f1> o") 'counsel-describe-symbol)
  (global-set-key (kbd "<f1> l") 'counsel-find-library)
  (global-set-key (kbd "<f2> i") 'counsel-info-lookup-symbol)
  (global-set-key (kbd "<f2> u") 'counsel-unicode-char)
  (global-set-key (kbd "C-c g") 'counsel-git)
  (global-set-key (kbd "C-c j") 'counsel-git-grep)
  (global-set-key (kbd "C-c k") 'counsel-ag)
  (global-set-key (kbd "C-x l") 'counsel-locate)
  (global-set-key (kbd "C-S-o") 'counsel-rhythmbox)
  (define-key minibuffer-local-map (kbd "C-r") 'counsel-minibuffer-history)
  ))

(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1)
  :custom ((doom-modeline-height 15)))

(add-to-list 'load-path "~/.emacs.d/lisp/emms")
(require 'emms-setup)
  (emms-all)
  (emms-default-players)
  (setq emms-source-file-default-directory "~/Music/") ;; Change to your music folder

(move-text-default-bindings)

(add-to-list 'load-path "~/.emacs.d/lisp/all-the-icons.el")

(when (display-graphic-p)
  (require 'all-the-icons))
;; or
(use-package all-the-icons
  :if (display-graphic-p))

(setq org-format-latex-options (plist-put org-format-latex-options :scale 2.0))

(set-fontset-font t 'symbol (font-spec :family "Noto Color Emoji") nil 'prepend)
(use-package emojify
  :config
  (when (member "Noto Color Emoji" (font-family-list))
    (set-fontset-font
     t 'symbol (font-spec :family "Noto Color Emoji") nil 'prepend))
  (setq emojify-display-style 'unicode)
  (setq emojify-emoji-styles '(unicode))
  (bind-key* (kbd "C-c .") #'emojify-insert-emoji)) ; override binding in any mode

(setq org-capture-templates
      '(("t" "Tarea" entry (file+headline "~/Documentos/org-mode/tareas.org" "Tareas")
         "* TODO %?\n  %i\n  %a")
        ("n" "Nota" entry (file+headline "~/Documentos/org-mode/notas.org" "Notas")
         "* %?\n  %i\n  %a")))

(global-set-key (kbd "C-c c") 'org-capture)

(set-frame-parameter (selected-frame) 'alpha '(100 . 100))
 (add-to-list 'default-frame-alist '(alpha . (100 . 100)))
